crates/miden-block-prover/src/tests/proposed_block_errors.rs

```diff
@@ -0,0 +1,628 @@
+      initial_state_commitment == account1.init_hash() &&
+      first_batch_id == batch0.id() &&
+      second_batch_id == batch1.id()
+    );
+
+    Ok(())
+}
+
+/// Tests that a missing account witness produces an error.
+#[test]
+fn proposed_block_fails_on_missing_account_witness() -> anyhow::Result<()> {
+    let TestSetup { mut chain, mut accounts, mut txs, .. } = setup_chain_without_auth(2);
+    let account0 = accounts.remove(&0).unwrap();
+    let tx0 = txs.remove(&0).unwrap();
+
+    let batch0 = generate_batch(&mut chain, vec![tx0]);
+
+    let batches = vec![batch0.clone()];
+
+    // This will not include the note inclusion proof for note0, because the note has not been added
+    // to the chain.
+    let mut block_inputs = chain.get_block_inputs(&batches);
+    block_inputs
+        .account_witnesses_mut()
+        .remove(&account0.id())
+        .expect("account witness should have been fetched");
+
+    let error = ProposedBlock::new(block_inputs, batches.clone()).unwrap_err();
+    assert_matches!(error, ProposedBlockError::MissingAccountWitness(account_id) if account_id == account0.id());
+
+    Ok(())
+}
+
+/// Tests that, given three transactions 0 -> 1 -> 2 which are executed against the same account and
+/// build on top of each other produce an error when tx 1 is missing from the block.
+#[test]
+fn proposed_block_fails_on_inconsistent_account_state_transition() -> anyhow::Result<()> {
+    // We need authentication because we're modifying accounts with the input notes.
+    let TestSetup { mut chain, mut accounts, .. } = setup_chain_with_auth(2);
+    let asset = generate_fungible_asset(
+        100,
+        AccountId::try_from(ACCOUNT_ID_FUNGIBLE_FAUCET_ON_CHAIN).unwrap(),
+    );
+
+    let account0 = accounts.remove(&0).unwrap();
+    let account1 = accounts.remove(&1).unwrap();
+
+    let note0 = generate_tracked_note_with_asset(&mut chain, account0.id(), account1.id(), asset);
+    let note1 = generate_tracked_note_with_asset(&mut chain, account0.id(), account1.id(), asset);
+    let note2 = generate_tracked_note_with_asset(&mut chain, account0.id(), account1.id(), asset);
+
+    // Add notes to the chain.
+    chain.seal_block(None);
+
+    // Create three transactions on the same account that build on top of each other.
+    // The MockChain only updates the account state when sealing a block, but we don't want the
+    // transactions to actually be added to the chain because of unintended side effects like spent
+    // nullifiers. So we create an alternative chain on which we generate the transactions, but
+    // then actually use the transactions on the original chain.
+    let mut alternative_chain = chain.clone();
+    let executed_tx0 = generate_executed_tx_with_authenticated_notes(
+        &mut alternative_chain,
+        account1.id(),
+        &[note0.id()],
+    );
+    alternative_chain.apply_executed_transaction(&executed_tx0);
+    alternative_chain.seal_block(None);
```

Comment:
@igamigo Is there an easier way to build transactions against the same account that build on top of each other but otherwise using transaction inputs from the MockChain?

Reply:
Thanks. Yeah in that case, I think the current approach may be the most concise. I think with the manual `TransactionContext` approach I'd have to also retrieve the transaction inputs from the chain "manually", so this would end up being more verbose I wager.